### PR TITLE
Win build 7.0.3 with cmake (static libs because of upstream issues)

### PR DESCRIPTION
**Pull request by Tobias-Fischer**
Checklist
* [x] Used a [personal fork of the feedstock to propose changes](https://conda-forge.org/docs/maintainer/updating_pkgs.html#forking-and-pull-requests)
* [x] Bumped the build number (if the version is unchanged)
* [ ] Reset the build number to `0` (if the version changed)
* [x] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [ ] Ensured the license file is being packaged.

### CODE REVIEW

**Comment by Tobias-Fischer** [comment]
@conda-forge-admin, please rerender

**Comment by Tobias-Fischer** [comment]
@conda-forge-admin, please rerender

**Comment by Tobias-Fischer** [comment]
@conda-forge-admin, please rerender

**Comment by Tobias-Fischer** [comment]
It seems as if upstream is not quite ready for windows yet. If someone would like to pick up from here, please let me know and I can provide access to my fork.

**Comment by Tobias-Fischer** [comment]
@conda-forge-admin, please rerender

**Comment by Tobias-Fischer** [comment]
Hi @traversaro, apologies for bothering you! I am trying to get scotch to build with Windows, and I got quite far. Indeed, the static build succeeds, but as you know, in conda-forge we prefer shared libs.

The problem is that by default no symbols are exported. Most issues can be fixed by `CMAKE_WINDOWS_EXPORT_ALL_SYMBOLS`. But there are a few static variables that need manual fixing (see build error in https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=668453&view=logs&jobId=a70f640f-cc53-5cd3-6cdc-236a1aa90802&j=a70f640f-cc53-5cd3-6cdc-236a1aa90802&t=f5d15007-a01c-5ad8-c9ce-4d519d3b275f). I thus use GenerateExportHeader to do some magic. However, I must be doing something wrong in https://github.com/Tobias-Fischer/scotch-feedstock/blob/win-build/recipe/0016-win-fix-exports.patch. It complains that:
```
D:\bld\scotch-pkgs_1678401110583\work\src\libscotch\bdgraph_bipart_st.c(84): error C2099: initializer is not a constant
D:\bld\scotch-pkgs_1678401110583\work\src\libscotch\bdgraph_bipart_st.c(99): error C2099: initializer is not a constant
D:\bld\scotch-pkgs_1678401110583\work\src\libscotch\bdgraph_bipart_st.c(104): error C2099: initializer is not a constant
D:\bld\scotch-pkgs_1678401110583\work\src\libscotch\bdgraph_bipart_st.c(184): error C2099: initializer is not a constant
```

A Google search tells me that this is expected, as C wants to know the initializers at compile time but when using declspec this won't be known until runtime: https://www.appsloveworld.com/c/100/118/windows-struct-initialization-error-c2099-initializer-is-not-a-constant

Do you have any idea how to resolve this issue?


Many thanks,
Tobi

**Comment by dalcinl** [comment]
> Do you have any idea how to resolve this issue?

Those used to develop code on POSIX are usually not aware of this issue on Windows. You cannot resolve this issue without actually changing the source code to move from compile-time initialization to runtime initialization. 
Could you please paste here how the C code reads? Are these global or function-local variaables?

**Comment by Tobias-Fischer** [comment]
The relevant source code looks as follows:
https://gitlab.inria.fr/scotch/scotch/-/blob/master/src/libscotch/bdgraph_bipart_st.c#L81-84
and
https://gitlab.inria.fr/scotch/scotch/-/blob/master/src/libscotch/parser.c#L81-82

Any help would be appreciated @dalcinl

**Comment by dalcinl** [comment]
Oh, boy... there is a lot of code to fix. First, let me clarify that I'm not an expert on the windows platform. That being said, one possible hard way to fix it is to replace all initializations by constants ( eg. all pointers like `&foo` by `NULL`). And then, at some point (e.g. the many `SCOTCH_*init()` routines) initialize properly all the globals.

Or maybe do something like


```C
#if defined(WINDOWS)
SomeType someglobalvar;
#else
SomeType someglobalvar = { the usual init code};
#endif
```
And then you add 
```c
#if defined(WINDOWS)
// some header include here, probably #include <windows.h>
BOOL WINAPI DllMain(HINSTANCE hinstDLL, DWORD fdwReason, LPVOID lpvReserved)
{
  if (fdwReason == DLL_PROCESS_ATTACH) {
    // initialize all globals here    
  }
  return TRUE;
}
#endif
```

Years ago, at a time Open MPI supported Windows, I had to modify mpi4py sources to make it work in that platform.

Maybe there are easier ways to do it. I would do some Googling first.

**Comment by traversaro** [comment]
> Do you have any idea how to resolve this issue?

I far from keyboard until Monday, I will check this on Monday!

**Comment by traversaro** [comment]
The first option fail with:
~~~
[242/555] Building C object src\libscotch\CMakeFiles\ptscotch.dir\bdgraph_bipart_ex.c.obj
[243/555] Building C object src\libscotch\CMakeFiles\ptscotch.dir\bdgraph_bipart_ml.c.obj
[244/555] Building C object src\libscotch\CMakeFiles\ptscotch.dir\bdgraph_bipart_sq.c.obj
[245/555] Building C object src\libscotch\CMakeFiles\ptscotch.dir\bdgraph_bipart_st.c.obj
FAILED: src/libscotch/CMakeFiles/ptscotch.dir/bdgraph_bipart_st.c.obj 
D:\bld\scotch-pkgs_1678725260680\_build_env\Library\bin\clang-cl.exe  /nologo -DCOMMON_OS_WINDOWS -DCOMMON_PTHREAD -DCOMMON_RANDOM_FIXED_SEED -DSCOTCH_PATCHLEVEL_NUM=3 -DSCOTCH_PTHREAD -DSCOTCH_RELEASE_NUM=0 -DSCOTCH_RENAME -DSCOTCH_VERSION_NUM=7 -Dptscotch_EXPORTS -Drestrict=__restrict -ID:\bld\scotch-pkgs_1678725260680\work\src\libscotch -ID:\bld\scotch-pkgs_1678725260680\work\build\src\libscotch -ID:\bld\scotch-pkgs_1678725260680\work\build\src\include -imsvcD:\bld\scotch-pkgs_1678725260680\_h_env\Library\include /DWIN32 /D_WINDOWS /W3 /MD /O2 /Ob2 /DNDEBUG -DSCOTCH_PTSCOTCH /showIncludes /Fosrc\libscotch\CMakeFiles\ptscotch.dir\bdgraph_bipart_st.c.obj /Fdsrc\libscotch\CMakeFiles\ptscotch.dir\ -c -- D:\bld\scotch-pkgs_1678725260680\work\src\libscotch\bdgraph_bipart_st.c
D:\bld\scotch-pkgs_1678725260680\work\src\libscotch\bdgraph_bipart_st.c(84,37):  error: initializer element is not a compile-time constant
} bdgraphbipartstdefaultbd = { { 3, &stratdummy } };
                                    ^~~~~~~~~~~
D:\bld\scotch-pkgs_1678725260680\work\src\libscotch\bdgraph_bipart_st.c(99,59):  error: initializer element is not a compile-time constant
} bdgraphbipartstdefaultml = { { 5, 1000, 2, 10000, 0.8L, &stratdummy, &stratdummy, &stratdummy} };
                                                          ^~~~~~~~~~~
D:\bld\scotch-pkgs_1678725260680\work\src\libscotch\bdgraph_bipart_st.c(104,34):  error: initializer element is not a compile-time constant
} bdgraphbipartstdefaultsq = { { &stratdummy } };
                                 ^~~~~~~~~~~
D:\bld\scotch-pkgs_1678725260680\work\src\libscotch\bdgraph_bipart_st.c(187,33):  error: initializer element is not a compile-time constant
                                (void *) &bgraphbipartststratab },
                                ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
4 errors generated.
[246/555] Building C object src\libscotch\CMakeFiles\ptscotch.dir\bdgraph_bipart_zr.c.obj
[247/555] Linking C shared library bin\scotch.dll
~~~ 

I am not sure if this happens because due to the use of flang we need to use an old clang on Windows. I will try the second option.

**Comment by traversaro** [comment]
The second option fails because in the scotch headers there is an attribute called class:
~~~
2023-03-13T17:58:37.8611350Z [249/555] Building CXX object src\libscotch\CMakeFiles\ptscotch.dir\bdgraph_bipart_st.c.obj
2023-03-13T17:58:37.8613116Z FAILED: src/libscotch/CMakeFiles/ptscotch.dir/bdgraph_bipart_st.c.obj 
2023-03-13T17:58:37.8616326Z D:\bld\scotch-pkgs_1678729921646\_build_env\Library\bin\clang-cl.exe  /nologo -TP -DCOMMON_OS_WINDOWS -DCOMMON_PTHREAD -DCOMMON_RANDOM_FIXED_SEED -DSCOTCH_PATCHLEVEL_NUM=3 -DSCOTCH_PTHREAD -DSCOTCH_RELEASE_NUM=0 -DSCOTCH_RENAME -DSCOTCH_VERSION_NUM=7 -Dptscotch_EXPORTS -Drestrict=__restrict -ID:\bld\scotch-pkgs_1678729921646\work\src\libscotch -ID:\bld\scotch-pkgs_1678729921646\work\build\src\libscotch -ID:\bld\scotch-pkgs_1678729921646\work\build\src\include -imsvcD:\bld\scotch-pkgs_1678729921646\_h_env\Library\include /DWIN32 /D_WINDOWS /W3 /GR /EHsc /MD /O2 /Ob2 /DNDEBUG -DSCOTCH_PTSCOTCH /showIncludes /Fosrc\libscotch\CMakeFiles\ptscotch.dir\bdgraph_bipart_st.c.obj /Fdsrc\libscotch\CMakeFiles\ptscotch.dir\ -c -- D:\bld\scotch-pkgs_1678729921646\work\src\libscotch\bdgraph_bipart_st.c
2023-03-13T17:58:37.8619490Z In file included from D:\bld\scotch-pkgs_1678729921646\work\src\libscotch\bdgraph_bipart_st.c:66:
2023-03-13T17:58:37.8621584Z D:\bld\scotch-pkgs_1678729921646\work\src\libscotch/arch.h(176,29):  error: expected member name or ';' after declaration specifiers
2023-03-13T17:58:37.8623687Z   const ArchClass *         class;                /*+ Pointer to architecture class         +*/
2023-03-13T17:58:37.8625514Z   ~~~~~~~~~~~~~~~           ^
2023-03-13T17:58:37.8627421Z D:\bld\scotch-pkgs_1678729921646\work\src\libscotch\bdgraph_bipart_st.c(112,56):  warning: ISO C++11 does not allow conversion from string literal to 'char *' [-Wwritable-strings]
2023-03-13T17:58:37.8629529Z                               { BDGRAPHBIPARTSTMETHBD, "b",  bdgraphBipartBd, &bdgraphbipartstdefaultbd },
2023-03-13T17:58:37.8636795Z                                                        ^
2023-03-13T17:58:37.8639502Z D:\bld\scotch-pkgs_1678729921646\work\src\libscotch\bdgraph_bipart_st.c(112,62):  error: cannot initialize a member subobject of type 'int (*)()' with an lvalue of type 'int (Bdgraph *const, const BdgraphBipartBdParam *const)' (aka 'int (Bdgraph_ *const, const BdgraphBipartBdParam_ *const)'): different number of parameters (0 vs 2)
2023-03-13T17:58:37.8642085Z                               { BDGRAPHBIPARTSTMETHBD, "b",  bdgraphBipartBd, &bdgraphbipartstdefaultbd },
2023-03-13T17:58:37.8644060Z                                                              ^~~~~~~~~~~~~~~
2023-03-13T17:58:37.8644676Z D:\bld\scotch-pkgs_1678729921646\work\src\libscotch/module.h(668,37):  note: expanded from macro 'bdgraphBipartBd'
2023-03-13T17:58:37.8645289Z #define bdgraphBipartBd             SCOTCH_NAME_INTERN (bdgraphBipartBd)
2023-03-13T17:58:37.8645989Z                                     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
2023-03-13T17:58:37.8646674Z D:\bld\scotch-pkgs_1678729921646\work\src\libscotch/module.h(247,37):  note: expanded from macro 'SCOTCH_NAME_INTERN'
2023-03-13T17:58:37.8647378Z #define SCOTCH_NAME_INTERN(f)       SCOTCH_NAME_MACRO3 (SCOTCH_NAME_PREFIX_INTERN,f,SCOTCH_NAME_SUFFIXC)
2023-03-13T17:58:38.0174032Z                                     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
2023-03-13T17:58:38.0321940Z D:\bld\scotch-pkgs_1678729921646\work\src\libscotch/module.h(246,37):  note: expanded from macro 'SCOTCH_NAME_MACRO3'
2023-03-13T17:58:38.0322998Z #define SCOTCH_NAME_MACRO3(p,n,s)   SCOTCH_NAME_GLUE3 (p,n,s)
2023-03-13T17:58:38.0323501Z                                     ^~~~~~~~~~~~~~~~~~~~~~~~~
2023-03-13T17:58:38.0324049Z D:\bld\scotch-pkgs_1678729921646\work\src\libscotch/module.h(244,37):  note: expanded from macro 'SCOTCH_NAME_GLUE3'
2023-03-13T17:58:38.0324688Z #define SCOTCH_NAME_GLUE3(p,n,s)    p##n##s
2023-03-13T17:58:38.0342547Z                                     ^~~~
2023-03-13T17:58:38.0343270Z <scratch space>(127,1):  note: expanded from here
2023-03-13T17:58:38.0343738Z _SCOTCHbdgraphBipartBd
~~~

I guess we are back to @dalcinl suggestions.

**Comment by Tobias-Fischer** [comment]
Thanks for highlighting the different options @traversaro @dalcinl! It seems like any of them will involve a considerable amount of work - unfortunately I won't have time to look into this; getting Windows to compile with the static builds already was a reasonable effort. I opened an issue upstream to make them aware of the conda-forge efforts: https://gitlab.inria.fr/scotch/scotch/-/issues/23

I agree with @traversaro's suggestion that a static library might be the best way forward for now until the remaining issues are fixed upstream. Do you have any objections/input @conda-forge/scotch? If not, I will revert back to the static build and mark the pull request as ready for review.

**Comment by dalcinl** [comment]
> I guess we are back to @dalcinl suggestions.

Now I realize I should have been more explicit/informative to prevent all this work you did. This is not a matter of C/C++ language or compiler versions or implementations. It is all about Window DLLs and the way DLL loading works in the Windows platform. Resistance is futile.

> https://gitlab.inria.fr/scotch/scotch/-/issues/23

Let see what they say. We would really need the help of the core developers to change the code, otherwise it is easy to mess up and endless to debug if all what we have are remote builds on a CI VM.

**Comment by Tobias-Fischer** [comment]
This is now ready for review @conda-forge/scotch

**Comment by Tobias-Fischer** [comment]
Kind reminder @conda-forge/scotch

File: recipe/build-scotch.sh [hunk 1]
```diff
@@ -51,6 +51,7 @@ cmake ${CMAKE_ARGS} \
   -D CMAKE_BUILD_TYPE=Release \
   -D CMAKE_SHARED_LINKER_FLAGS="$shared_flags" \
   -D CMAKE_INSTALL_PREFIX=$PREFIX \
+  -D CMAKE_INSTALL_RPATH=${PREFIX}/lib \
```
Comment (dalcinl) [inline recipe/build-scotch.sh:54]: This one worries me a little. conda is supposed to fix the rpath. Do we really need it or is it a leftover of the Windows nightmare you went through?

Comment (Tobias-Fischer) [reply recipe/build-scotch.sh]: Good catch @dalcinl. Removed this flag, it was indeed just a leftover.

**Comment by dalcinl** [comment]
@Tobias-Fischer Could you please confirm whether this PR is ready to merge?

**Pull request merged**